### PR TITLE
Add Redis Pub/Sub and Docker events

### PR DESCRIPTION
**Pull request by mishan**
- Resolves issue #4 and partially resolves issue #3 

@TysonAndre

### CODE REVIEW

**Comment by mishan** [comment]
Most of this is working code I took back from ecssd_agent heh

File: redecs_agent/redecs_agent.go [hunk 1]
```diff
@@ -44,10 +133,24 @@ func logErrorNoFatal(err error) {
...
 	// write {timestamp => SERVICENAME_IP} to ecssd:service_pings
-	redisClient.ZAdd("redecs:service_pings", redis.Z{float64(time.Now().Unix()), serviceName + "_" + configuration.LocalIp})
+	redisClient.ZAdd("redecs:service_pings", redis.Z{float64(time.Now().Unix()), value})
+	// notify the channel that this service has been reported active
+	redisClient.Publish("redecs:service_channel", "+"+value)
+	return nil
+}
+
+func reportServiceInactive(serviceName string) error {
+	log.Debug("Reporting " + serviceName + " as active with IP " + configuration.LocalIp)
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go]: nit: debugf

File: redecs_agent/redecs_agent.go [hunk 2]
```diff
@@ -44,10 +133,24 @@ func logErrorNoFatal(err error) {
 var dockerClient *docker.Client
 var redisClient *redis.Client
 
-func reportServiceActive(serviceName string) {
+func reportServiceActive(serviceName string) error {
 	log.Debug("Reporting " + serviceName + " as active with IP " + configuration.LocalIp)
+	value := serviceName + "_" + configuration.LocalIp
 	// write {timestamp => SERVICENAME_IP} to ecssd:service_pings
-	redisClient.ZAdd("redecs:service_pings", redis.Z{float64(time.Now().Unix()), serviceName + "_" + configuration.LocalIp})
+	redisClient.ZAdd("redecs:service_pings", redis.Z{float64(time.Now().Unix()), value})
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go]: `err := redisClient.Zadd(...).Err()`

https://godoc.org/gopkg.in/redis.v5#IntCmd.Err

File: redecs_agent/redecs_agent.go [hunk 3]
```diff
@@ -20,6 +20,95 @@ import (
...
+
+type dockerRouter struct {
+	handlers      map[string][]handler
+	dockerClient  *docker.Client
+	listener      chan *docker.APIEvents
+	workers       chan *worker
+	workerTimeout time.Duration
+}
+
+func dockerEventsRouter(bufferSize int, workerPoolSize int, dockerClient *docker.Client,
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go]: dockerRouter can be moved into its own file or subpackage (everything within a folder is in the same package)

File: redecs_agent/redecs_agent.go [hunk 4]
```diff
@@ -44,10 +133,24 @@ func logErrorNoFatal(err error) {
 var dockerClient *docker.Client
 var redisClient *redis.Client
 
-func reportServiceActive(serviceName string) {
+func reportServiceActive(serviceName string) error {
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go:48]: Add one line comments for the long functions in the format `// reportServiceActive reports this service as active in redis`

File: redecs_agent/redecs_agent.go [hunk 5]
```diff
@@ -44,10 +133,24 @@ func logErrorNoFatal(err error) {
...
+	// notify the channel that this service has been reported active
+	redisClient.Publish("redecs:service_channel", "+"+value)
+	return nil
+}
+
+func reportServiceInactive(serviceName string) error {
+	log.Debug("Reporting " + serviceName + " as active with IP " + configuration.LocalIp)
+	value := serviceName + "_" + configuration.LocalIp
+	// remove SERVICENAME_IP from ecssd:service_pings
+	redisClient.ZRem("redecs:service_pings", value)
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go]: `err := redisClient.ZRem().Err()`

File: redecs_agent/redecs_agent.go [hunk 6]
```diff
@@ -20,6 +20,95 @@ import (
...
+			if err := handler.Handle(event); err != nil {
+				log.Errorf("Error processing event %#v. Error: %v", event, err)
+			}
+		}
+	}
+}
+
+type dockerHandler struct {
+	handlerFunc func(event *docker.APIEvents) error
+}
```
Comment (TysonAndre) [inline redecs_agent/redecs_agent.go]: can say `var _ handler = &dockerHandler{}` to statically assert dockerHandler implements handler.

File: red53_agent/red53_agent.go [hunk 7]
```diff
@@ -119,6 +121,7 @@ func fetchActiveServices() {
 	var servicePings []string
 	var sum int
 	var err error
+	modifyMutex.Lock()
```
Comment (TysonAndre) [inline red53_agent/red53_agent.go:124]: The typical pattern is

``` go
m.Lock()
defer m.Unlock()
```

This ensures that the mutex will be unlocked even after a `panic` , and avoids forgetting when inserting an early `return`

File: red53_agent/red53_agent.go [hunk 8]
```diff
@@ -204,12 +207,28 @@ func main() {
 		log.Fatal(http.ListenAndServe(":8080", nil))
 	}()
 
+	serviceChanges, err := redisClient.Subscribe("redecs:service_channel")
+	logErrorAndFail(err)
+	defer serviceChanges.Close()
+
 	// check regularly, specified by checkInterval
 	ticker := time.NewTicker(checkInterval)
```
Comment (TysonAndre) [inline red53_agent/red53_agent.go]: ticker can be declared in into the below goroutine

File: redecs_agent/redecs_agent.go (hunk 3 above)
Comment (mishan) [reply redecs_agent/redecs_agent.go]: done

**Review by TysonAndre — COMMENTED** [review]
Forgot to submit 2 comments  earlier:

(Other refactoring seems fine.)

File: red53_agent/red53_agent.go (hunk 7 above)
Comment (mishan) [reply red53_agent/red53_agent.go:124]: Fixed, thanks!

**Pull request merged**